public/app/features/dimensions/editors/ResourcePickerPopover.tsx

```diff
@@ -0,0 +1,273 @@
+import { ResourceFolderName } from '..';
+
+interface Props {
+  value?: string; //img/icons/unicons/0-plus.svg
+  onChange: (value?: string) => void;
+  mediaType: 'icon' | 'image';
+  folderName: ResourceFolderName;
+}
+
+export interface ResourceItem {
+  label: string;
+  value: string; // includes folder
+  search: string;
+  imgUrl: string;
+}
+
+type PickerType = 'folder' | 'url';
+
+const getFolders = (mediaType: 'icon' | 'image') => {
+  if (mediaType === 'icon') {
+    return [ResourceFolderName.Icon, ResourceFolderName.IOT, ResourceFolderName.Marker];
+  } else {
+    return [ResourceFolderName.BG];
+  }
+};
+
+const getFolderIfExists = (folders: Array<SelectableValue<string>>, path: string) => {
+  return folders.find((folder) => path.startsWith(folder.value!)) ?? folders[0];
+};
+
+export const ResourcePickerPopover = (props: Props) => {
+  const { value, onChange, mediaType, folderName } = props;
+
+  const folders = getFolders(mediaType).map((v) => ({
+    label: v,
+    value: v,
+  }));
+
+  const [currentFolder, setCurrentFolder] = useState<SelectableValue<string>>(
+    getFolderIfExists(folders, value?.length ? value : folderName)
+  );
+  const [directoryIndex, setDirectoryIndex] = useState<ResourceItem[]>([]);
+  const [filteredIndex, setFilteredIndex] = useState<ResourceItem[]>([]);
+  useEffect(() => {
+    // we don't want to load everything before picking a folder
+    const folder = currentFolder?.value;
+    if (folder) {
+      const filter =
+        mediaType === 'icon'
+          ? (item: FileElement) => item.name.endsWith('.svg')
+          : (item: FileElement) => item.name.endsWith('.png') || item.name.endsWith('.gif');
+
+      getDatasourceSrv()
+        .get('-- Grafana --')
+        .then((ds) => {
+          (ds as GrafanaDatasource).listFiles(folder).subscribe({
+            next: (frame) => {
+              const cards: ResourceItem[] = [];
+              frame.forEach((item) => {
+                if (filter(item)) {
+                  const idx = item.name.lastIndexOf('.');
+                  cards.push({
+                    value: `${folder}/${item.name}`,
+                    label: item.name,
+                    search: (idx ? item.name.substr(0, idx) : item.name).toLowerCase(),
+                    imgUrl: `public/${folder}/${item.name}`,
+                  });
+                }
+              });
+              setDirectoryIndex(cards);
+              setFilteredIndex(cards);
+            },
+          });
+        });
+    }
+  }, [mediaType, currentFolder]);
+
+  const [searchQuery, setSearchQuery] = useState<string>();
+  const onChangeSearch = (query: string) => {
+    if (query) {
+      query = query.toLowerCase();
+      setFilteredIndex(directoryIndex.filter((card) => card.search.includes(query)));
+    } else {
+      setFilteredIndex(directoryIndex);
+    }
+  };
+
+  const [newValue, setNewValue] = useState<string>(value ?? '');
```

Comment:
A good practice with hooks is to defined them at the top level of the component. Otherwise you're in a reisk of introducing i.e. early exit before the hook.

Reply:
Sounds good. In this case I was trying to group similar logic which is one of the benefits of hooks over classes. In general does eng at Grafana prefer grouping all hooks at top of every file vs. grouping similar logic?